pyiron/atomistics/structure/atoms.py

```diff
@@ -1058,115 +1058,208 @@ def analyse_phonopy_equivalent_atoms(atoms):
+        """
+        Writes a PDB-formatted line to represent an atom.
+
+        Args:
+            num (int): Atomic index.
+            species (str): Elemental species.
+            group (str):
+            num2 (int): An "alternate" index. (Don't ask me...)
+            coords (list/ndarray[float]: X, Y, and Z coordinates of the atom.
+            occupancy (float): PDB occupancy parameter.
+            temperature_factor (float): PDB temperature factor parameter.
+
+        Returns:
+            (str): The line defining an atom in PDB format
+
+        Warnings:
+            * The [PDB docs](https://www.cgl.ucsf.edu/chimera/docs/UsersGuide/tutorials/pdbintro.html) indicate that
+                the xyz coordinates might need to be in some sort of orthogonal basis. If you have weird behaviour,
+                this might be a good place to investigate.
+        """
         x, y, z = coords
-        return 'ATOM {:>6} {:>4} {:>4} {:>5} {:10.3f} {:7.3f} {:7.3f} {:5.2f} {:5.2f} {:>11} \n'.format(num, species, group, num2, x, y, z, c0, c1, species)
+        return 'ATOM {:>6} {:>4} {:>4} {:>5} {:10.3f} {:7.3f} {:7.3f} {:5.2f} {:5.2f} {:>11} \n'.format(
+            num, species, group, num2, x, y, z, occupancy, temperature_factor, species)
     
     def _ngl_write_structure(self, elements, positions, cell, scalar_field=None):
+        """
+        Turns structure information into a NGLView-readable protein-database-formatted string.
+
+        Args:
+            elements (ndarray/list): Element symbol for each atom.
+            positions (ndarray/list): Vector of Cartesian atom positions.
+            cell (ndarray/list): Simulation cell Bravais matrix.
+            scalar_field (ndarray/list): Per-atom vector of floats used to re-color the atoms. (Default is None, use
+                NGLView coloring scheme.)
+
+        Returns:
+            (str): The PDB-formatted representation of the structure.
+        """
         from ase.geometry import cell_to_cellpar, cellpar_to_cell
         cellpar = cell_to_cellpar(cell)
         exportedcell = cellpar_to_cell(cellpar)
         rotation = np.linalg.solve(cell, exportedcell)
   
-        pdb_str = self._ngl_write_cell(cellpar[0], cellpar[1], cellpar[2], cellpar[3], cellpar[4], cellpar[5])
+        pdb_str = self._ngl_write_cell(*cellpar)
         pdb_str += 'MODEL     1\n'
+
         if scalar_field is None:
             scalar_field = np.ones(len(positions))
         else:
-            scalar_field = (scalar_field-np.min(scalar_field))/(np.max(scalar_field)-np.min(scalar_field))
+            scalar_field = (scalar_field - np.amin(scalar_field)) / np.ptp(scalar_field)
+
         for i, p in enumerate(positions):
             if rotation is not None:
                 p = p.dot(rotation)
                 
-            pdb_str += self._ngl_write_atom(i, elements[i], group=elements[i], num2=i, coords=p, c0=scalar_field[i], c1=0.0)
+            pdb_str += self._ngl_write_atom(i, elements[i],
```

Comment:
Is it possible to avoiding looping over the atoms altogether? For example, the positions can be formated into a string format of your choice using some nice numpy routines

`np.array2string(np.char.mod(('%10.3f', '%7.3f', '%7.3f'), self.positions)).replace("]", "").replace("[", "")`

Reply:
Yes, I'm sure it's possible, but I'm actually going to skip this one. I think any structure big enough makes this for-loop give a noticeable performance hit is a structure which will cause problems anyhow. In the meantime, I find the for-loop much more readable than a string-hacked array. If we run into performance issues we can change it later, but it's always been this way so I think we're safe ;)
(Note that I do get some performance hit on my new re-coloring which calls a separate `add_spacefill` for each atom, but it's about 0.5-1s for a 300 atom cell so I'll live with it for now.)